### ISSUE TITLE
Feature Request: Option to hide zero values in legend and tooltip

### ISSUE THREAD

**Issue by jmcle** [grafana/grafana#1381]
The new options to hide null values in the legend and 'all series' tooltip are excellent but I run into a problem when using them with stacked bar graphs.

With stacked graphs we need to set 'null as zero' in order to get the stacking to work properly but this negatively impacts the ability to hide null values in the legend and tooltip as null values will show up as zero instead of null.

I'm not sure if there's a better way to address this but one option would be to add an option to 'hide zero values' in addition to the existing 'hide null values'. Another option would be to make the 'hide value' option a parameter which you could set to any numeric value. This would be similar to what the 'removeBelowValue' function does but in this case it would affect the display of legends and tooltips.

**Comment by fzaire** [comment]
I have the same issue and I end up with 80% of the metrics in my legend with 0 value that i need to hide.
If more people are facing the same issue, we could try to commit this feature.

**Comment by torkelo** [comment]
cant you filter out the zero metrics/series in your query?

**Comment by linar-jether** [comment]
+1 for this issue, as we can't filter out zeroes without breaking stacks, i.e 
![image](https://cloud.githubusercontent.com/assets/11908736/10841685/57bd3704-7ef4-11e5-803c-e71854815769.png)

**Comment by agourlay** [comment]
+1 for this issue.

**Comment by apakulov** [comment]
+1

**Comment by utkarshcmu** [comment]
+1

**Comment by digrich** [comment]
+1

**Comment by george-vieira** [comment]
I have the same issue, my metrics shows traffic by IP and they spike in and out so at any point in time there could be values or 0. The tooltip shows 100 IPs which is overall but I only care of the point in time I'm hovering over.

**Comment by utkarshcmu** [comment]
The above PR only hides the zero values if all datapoints are 0. Even if a single datapoint is non-zero on the graph for a given time range, it will not hide the metric.

**Comment by Pryz** [comment]
+1

It could be nice to specify a pattern. Like "hide if < 0" or "hide if = 0" etc

**Comment by utkarshcmu** [comment]
As an example, Opentsdb cannot remove zeros in query itself, so it would be great if Grafana can eliminate them from Legend and Tooltip. And looks like a lot of people wants this feature to be there in Grafana.

@torkelo Are these +1s enough for #3336 to get merged? Please let me know if any refinement is needed in the PR #3336.

**Comment by dacr** [comment]
+1

with a pattern such as "hide when no value" (in current time range)

**Comment by torkelo** [comment]
![image](https://cloud.githubusercontent.com/assets/10999/11501575/8f5950f8-9835-11e5-9587-1956169ba481.png)

**Comment by linar-jether** [comment]
with hide a series only when all of the values are null/zero? what about the tooltip when only part of a series is null/zero and we don't want to show empty values? (see image above)

**Comment by torkelo** [comment]
These options apply to the tooltip as well. They are only applied then all the values in the series are null or zero

**Comment by george-vieira** [comment]
if the whole series were to be null or 0 then that can be added as part of the query. My problem is where the series have values > 0 but most times at any part of the graph, my tooltip shows every IP address regardless of value or 0. Image describes it best.
So I don't mind the legend but tooltip showing only at the point in time, values that are not null or over 0 in value.

![tooltip](https://cloud.githubusercontent.com/assets/13330803/11524372/747990e2-991e-11e5-8f37-54905cbeb46a.png)

**Comment by roysbike** [comment]
george-vieira +1 . Please add function. If value =0 hide

**Comment by rgombash** [comment]
+1

**Comment by torkelo** [comment]
@rgombash ýou do not need to +1 as this is already implemented and in latest 2.6 release

**Comment by slava-vishnyakov** [comment]
@torkelo I'm running 2.6.0, but I'm still seeing 0 in tooltips even with 2 options "on". 

Frankly it would be really useful to see only non-zero values. We have graphs with hundreds of series, but only 1 or 2 are actually non-zero at each time. 

![image](https://cloud.githubusercontent.com/assets/817931/13301064/275a8b68-db54-11e5-84e2-b819f66c00ae.png)

Often things are much worse:

![image](https://cloud.githubusercontent.com/assets/817931/13301157/9460bfe8-db54-11e5-97d2-ec993dc781b2.png)

Thanks for Grafana! Otherwise it's awesome.

**Comment by torkelo** [comment]
@slava-vishnyakov the feature only hides series where are all values are zero. 

if you have a lot of series where only some have values, can you not filter the metrics in your query?

**Comment by slava-vishnyakov** [comment]
@torkelo No, sadly not. Imagine hundreds of microservices that occasionally get 500 Internal error. They might get it only once every 10 minutes. So, now I'm looking at a graph that has a few huge spikes and lot of "dots". Those "dots" make those series "non-zero". I can't also predict which services are going to be failing and filter them in query.

But anyway, seeing those 0s at tooltip is IMO unhelpful. Basically, it says something like, "ok here's what you don't see at this point". For me it seems that charts might benefit from not showing zero values at tooltips even by default.

**Comment by george-vieira** [comment]
@slava-vishnyakov: that's the same issue I have as my previous pic shows. My case is a large list of IPs showing traffic data. 1 host in the series speaks for 1 second and it's on the whole time line and tooltip. So when a bunch of are all chatty at differnt times and 0 on others and then there's a spike I want to look at, all of the IPs are shown on the tooltip even though on the exact time of mouse over inspect, I get a huge tooltip.
I can't filter the series because there aren't always 0 through the whole time range query but occassionally appear in the series here and there but for that range, possibly 100 are listed all at different spots of the time line.

**Comment by scraly** [comment]
Same issue here.
We need to display in tooltip only non-null non-0 values.
Thanks.

![grafana_0](https://cloud.githubusercontent.com/assets/2543381/14529470/f74551d8-0254-11e6-9da2-a56308f200d1.png)

![grafana_legend](https://cloud.githubusercontent.com/assets/2543381/14529513/3664dfdc-0255-11e6-92be-cf690849deb5.png)

**Comment by seabadger** [comment]
+1. 
I have a large number of data series representing error occurrences for different services. A few series (representing services with on-going issues) are non zero across the time range. The rest are either zero at all times, or have an occasional blip or two. 
The tooltip shows all values that had even a single blip in the time range, regardless of whether the mouse is hovering over the blip or not. 

What'd be really useful for me is if the tooltip were a bit more dynamic, and could be configured to show only non-zero values at a given point (thus, the set of rows in the tooltip would be changing as the mouse moves around the graph).

In other words: 
the tooltip represents just the part of the graph under the mouse cursor, and should be able to reflect only what's shown there (configurably, obviously, since not everyone needs/wants this).

This is in contrast with the legend, which represents the entire graph, so showing all series with non-zero values in the time range makes a lot more sense there.

**Comment by seabadger** [comment]
just reazlied that (1) this issue is closed, and my issue is also covered by #4623

**Comment by jveldboom** [comment]
For anyone else trying to hide the metrics from the tooltip & legend as well as the graph itself, you can use both the Legend option "Hide series > With only zeros" as well as adjust the Axes Y-Min to `.1`

![image](https://user-images.githubusercontent.com/303202/32138147-276f05e0-bbfc-11e7-8aff-e4e5edb5f7e7.png)

![image](https://user-images.githubusercontent.com/303202/32138149-3291b120-bbfc-11e7-948f-e0aa8d9181c6.png)

Will turn this:
![image](https://user-images.githubusercontent.com/303202/32138159-50372174-bbfc-11e7-851a-a5e6b6f92f0b.png)

To this:
![image](https://user-images.githubusercontent.com/303202/32138179-6a2ecc4e-bbfc-11e7-9456-8de26cace1c5.png)

**Comment by lambdaq** [comment]
@jveldboom Hover Tooltip still has zero values.

**Comment by AndySweetIQ** [comment]
Same issue here. I have a graph where some important metrics occasionally will spike, so I can't filter them out. So, because they have values anywhere in the time range (monthly, for reporting), they basically pollute the tooltip **by increasing the size of it's rectangle**, forcing us to constantly trying to move the mouse cursor around to see what's under the tooltip.

I do sort the entires in the tooltip by decreasing order, but it only helps a little.
Hopefully this gets worked on someday!

Other than that, Grafana has been a god sent for us.

**Comment by dangerrockstar** [comment]
hi there 
i have also face same problem

select count(average_price) from test group by time(1d)
then he collate lots of '0' zero,so how to i ignore that zero data ?

**Comment by cas--** [comment]
For reference tooltip hide zero ticket is #4623

**Comment by slava-vishnyakov** [comment]
#4623 was opened more than a year later than this one... Wow, it has been 3 years already...

**Comment by calhoun137** [comment]
Here is a quick fix.  Try adding display none on zero's like this:

```
        var trowEnter = tbodyEnter.selectAll("tr")
                .data(function(p) { return p.series})
                .enter()
                .append("tr")
                .classed("highlight", function(p) { 
                    if (p.data.y === '0') {
                        this.style.display = "none";                    
                    } 

                    return p.highlight
                });
```

**Comment by Arch1tect** [comment]
+1

**Comment by satishdotpatel** [comment]
+1 we need this feature 👍

**Comment by ageis** [comment]
this still open? 😒

**Comment by halsafar** [comment]
Can't believe I just read through 2 years of comments on this and it is still not implemented.
+1...

**Comment by HardPeach** [comment]
+1 This feature is still not available out of the blue.

**Comment by 1337blue** [comment]
+1 BTW chronograf does this oob

**Comment by ptchau2003** [comment]
+1

**Comment by gilsand** [comment]
+1

**Comment by herlon214** [comment]
+1

**Comment by jeremie-H** [comment]
+1
We should at least reopen this issue. It's not fixed. more than 4 years ago :'(
poke @torkelo

**Comment by torkelo** [comment]
Not sure what you mean this feature has existed for some time and works fine. 

if you toggle this option in the legend options it will hide series with only zeros in legend and tooltip 
<img width="785" alt="Screen Shot 2019-10-16 at 13 34 07" src="https://user-images.githubusercontent.com/10999/66915600-b1d6be00-f019-11e9-8d63-eb9a694cc703.png">

**Comment by jeremie-H** [comment]
Hi @torkelo and thanks for your answer !  

The option "Hide series" works but it's not the wanted feature here.  
The option "Hide series" will deactivate all series with only zeros for example, and it's great to keep it.  
 
But if a serie has only one peak of data, and the rest is zero in the time range, the serie will appear and the feature request here is on **hover tooltip** to deactivate all data with *0* values.  

Below you can see that, in the time range, we've got a first peak at 09:07 with number of series.  
And at 09:14 you can see other series, if i let my mouse with the hoover tooltip, i've got numbers of series with series (even with **Hide series with only zeros** activate)

![grafana2](https://user-images.githubusercontent.com/11060749/66922869-2f092f80-f028-11e9-9443-bf87347cc02f.png)

And then with a great number of series, hoover toolip is sorted by decreasing ... I even can't read the series with values, because tooltip is out of range in the screen (and scroll is up)
![grafana3](https://user-images.githubusercontent.com/11060749/66922881-34ff1080-f028-11e9-9830-a3d60aa4a95c.png)

**Comment by torkelo** [comment]
That is unrelated to the feature request your commenting on

**Comment by jeremie-H** [comment]
TITLE :
> Feature Request: Option to hide zero values in legend **and tooltip**

Am I missing something or perhaps another issue on it ?

**Comment by torkelo** [comment]
This feature was closed as it only aksed for hiding series with all zeros in legend and tooltip, implemented and shipped in v2.6

**Comment by jeremie-H** [comment]
But, zeros values are still displayed in tooltip as you can see.

I'm just a humble developer, I search on internet, i felt here, i can see more than 40 comments, a lot of people need this feature and a lot of +1 for 4 years...

I search more and find this issue which is exactly the problem I encounter..  https://github.com/grafana/grafana/issues/3094

But You've marked as a duplicate here ...

Can you reconsider your position ?

**Comment by jmcle** [comment]
I can confirm that my original request was never fully resolved and Jeremie and the multiple other people who have chimed in about this are correct. It would be great if the request could be acknowledged and guidance provided in terms of how to proceed here.

**Comment by torkelo** [comment]
From the issue description:

> I'm not sure if there's a better way to address this but one option would be to add an option to 'hide zero values' in addition to the existing 'hide null values'. 

Which is exactly what the PR did that closed this issue

**Comment by torkelo** [comment]
I am not at all opposed to what your asking but please comment on a new feature request or maybe this related issue https://github.com/grafana/grafana/issues/6946

**Comment by d3noob** [comment]
For those (such as I was) looking for an option to remove zero values from a series when this represents an innacurate reading (imagine an atmospheric pressure reading which will sometimes return 0 when typical values will be from 980 - 1030), the resolution that I found easiest to implement was to add a qualifier to the query that rejected zero values. E.g; `weather_barometer_mb != 0`.
This had the desired effect of changing ...
![image](https://user-images.githubusercontent.com/3151700/73583485-69bfa680-44f7-11ea-9454-a12e992e0858.png)
to...
![image](https://user-images.githubusercontent.com/3151700/73583517-8bb92900-44f7-11ea-8090-60e4198d5583.png)
I realise that this will not remove zero value metrics from the tooltips, but for those like myself who were following this thread looking for the solution just explained, perhas this will assist.

**Comment by bl4ckmesa** [comment]
For my part, changing the order in tool tips was good enough to get along for now.  Under Visualization -> Hover Tooltip -> Sort Order I set it to Decreasing.  The tooltip still looks ugly as sin but at least the values go on top and I can see the numbers.

**Comment by tomlut** [comment]
> I found easiest to implement was to add a qualifier to the query that rejected zero values. E.g; `weather_barometer_mb != 0`

What was the actual query you used to implement this?

I'm having trouble with the syntax.

e.g. how would I reject 0 values from this:

`SELECT min("value") FROM "MB" WHERE ("entity_id" = 'db_size') AND $timeFilter GROUP BY time(1d) fill(null)`

**Comment by d3noob** [comment]
From @tomlut 

> What was the actual query you used to implement this?

Actually that is the query. The data is pre-normalised which makes the process look really simple.


> e.g. how would I reject 0 values from this:
> 
> `SELECT min("value") FROM "MB" WHERE ("entity_id" = 'db_size') AND $timeFilter GROUP BY time(1d) fill(null)`

Apologies, I would be learning how to do it at the same time you would be. The best advice I would give is to experiment liberally with brackets and google.

**Comment by anacrolix** [comment]
I don't follow why this issue is closed. There are several related issues and PRs that show this is an ongoing issue.

**Comment by anikpati89** [comment]
This issue is not yet fixed. Please reopen the ticket

**Comment by jmfiola** [comment]
please reopen

**Comment by jeremie-H** [comment]
if you fall here, dont loose your time, switch to influxdb/chronograf or anything else and giveup with grafana ;-)
this issue, 7 years old, is a bad joke and will never be fixed !

it's the best advice i can give here :)

**Comment by AnnaWojciechowska** [comment]
sadly customizing the dashboard to make them user-friendly is a huge time sink and usually not much doable.  See the example below (loosing hope)
![Screenshot from 2023-03-14 10-37-04](https://user-images.githubusercontent.com/10974742/224960165-77287bed-e1a8-46e8-8b5a-4716508f6f86.png)
t

**Comment by Xasin** [comment]
Let me add in my own bit of fuff:

I seco... Err... Fifteeth-or-so the opinion that this issue is *not* resolved and that this issue shall be reopened.
It should be a fairly easy thing to filter out all Legend-Entries that, for the currently-being-shown-legend, have a null or 0 value.
As was stated many times, certain workloads with a large number of series that only occasionally spike and are mostly zero will lead to an unreadable legend that is hard to parse.

My use case, monitoring when certain process CPU usages spike, looks like this:
![image](https://user-images.githubusercontent.com/4147002/233868098-cab6ed95-be33-4f3d-a603-88451be46f3d.png)

And I think it's quite silly to have only 5% useful information displayed in the legend.

**Comment by leeoniya** [comment]
this is fixed in 9.5 (see https://github.com/grafana/grafana/pull/64477)

**Comment by lucumt** [comment]
> if you fall here, dont loose your time, switch to influxdb/chronograf or anything else and giveup with grafana ;-) this issue, 7 years old, is a bad joke and will never be fixed !
> 
> it's the best advice i can give here :)

Two years has gone,this issue sames still exsits!

**Comment by melroy89** [comment]
I still see zero values in my graph, tooltip and legend. Despite using "Fields with values" -> "All Zeros". And trying to hide: Tooltip, Viz + Legend:

![image](https://github.com/user-attachments/assets/4b6c53d0-d04f-4378-a5b8-5c82d9e99e10)

**Comment by leeoniya** [comment]
this was implemented for Tooltip in https://github.com/grafana/grafana/pull/97986
